tensorflow/compiler/xla/service/slice_sinker_test.cc

```diff
@@ -0,0 +1,551 @@
+#include "tensorflow/compiler/xla/service/hlo_computation.h"
+#include "tensorflow/compiler/xla/service/hlo_instruction.h"
+#include "tensorflow/compiler/xla/service/hlo_module.h"
+#include "tensorflow/compiler/xla/service/hlo_opcode.h"
+#include "tensorflow/compiler/xla/service/hlo_parser.h"
+#include "tensorflow/compiler/xla/service/pattern_matcher.h"
+#include "tensorflow/compiler/xla/service/pattern_matcher_gmock.h"
+#include "tensorflow/compiler/xla/shape_util.h"
+#include "tensorflow/compiler/xla/tests/hlo_test_base.h"
+#include "tensorflow/compiler/xla/types.h"
+#include "tensorflow/compiler/xla/xla_data.pb.h"
+#include "tensorflow/core/lib/core/status_test_util.h"
+#include "tensorflow/core/platform/types.h"
+
+namespace xla {
+namespace {
+
+using ::testing::ElementsAre;
+namespace m = match;
+
+class SliceSinkerTest : public HloTestBase {
+};
+
+TEST_F(SliceSinkerTest, Basic) {
+  const char* kModuleStr = R"(
+    HloModule m
+    test {
+      p0 = pred[8,9] parameter(0)
+      p1 = f32[8,9] parameter(1)
+      p2 = f32[8,9] parameter(2)
+      s00 = pred[2,9] slice(pred[8,9] p0), slice={[0:2], [0:9]}
+      s01 = pred[6,9] slice(pred[8,9] p0), slice={[2:8], [0:9]}
+      s10 = f32[2,9] slice(f32[8,9] p1), slice={[0:2], [0:9]}
+      s11 = f32[6,9] slice(f32[8,9] p1), slice={[2:8], [0:9]}
+      s20 = f32[2,9] slice(f32[8,9] p2), slice={[0:2], [0:9]}
+      s21 = f32[6,9] slice(f32[8,9] p2), slice={[2:8], [0:9]}
+      sel0 = f32[2,9] select(pred[2,9] s00, f32[2,9] s10, f32[2,9] s20)
+      sel1 = f32[6,9] select(pred[6,9] s01, f32[6,9] s11, f32[6,9] s21)
+      ROOT tuple = (f32[2,9], f32[6,9]) tuple(sel0, sel1)
+    }
+  )";
+  TF_ASSERT_OK_AND_ASSIGN(auto module,
+                          ParseAndReturnVerifiedModule(kModuleStr));
+  SliceSinker slice_sinker;
+  TF_ASSERT_OK_AND_ASSIGN(bool result,
+                          RunHloPass(&slice_sinker, module.get()));
+  EXPECT_TRUE(result);
+  HloDCE dce;
+  TF_ASSERT_OK_AND_ASSIGN(result,
+                          RunHloPass(&dce, module.get()));
+  EXPECT_TRUE(result);
+  EXPECT_EQ(7, module->entry_computation()->instruction_count());
+  HloInstruction* inst = module->entry_computation()->root_instruction();
+  EXPECT_THAT(inst, GmockMatch(m::Tuple(
+      m::Slice(m::Select(m::Parameter(0), m::Parameter(1), m::Parameter(2))),
+      m::Slice(m::Select(m::Parameter(0), m::Parameter(1), m::Parameter(2))))));
+  HloInstruction* slice0 = inst->mutable_operand(0);
+  EXPECT_EQ(slice0->slice_starts(), std::vector<int64>({0, 0}));
+  EXPECT_EQ(slice0->slice_limits(), std::vector<int64>({2, 9}));
+  EXPECT_EQ(slice0->slice_strides(), std::vector<int64>({1, 1}));
+  HloInstruction* slice1 = inst->mutable_operand(1);
+  EXPECT_EQ(slice1->slice_starts(), std::vector<int64>({2, 0}));
+  EXPECT_EQ(slice1->slice_limits(), std::vector<int64>({8, 9}));
+  EXPECT_EQ(slice1->slice_strides(), std::vector<int64>({1, 1}));
+}
+
+TEST_F(SliceSinkerTest, PartialButOverlappingSlicesTransformed) {
```

Comment:
This name doesn't look consistent with the test, let's rename it to OverlappingPartialSlicesBeneficial